### PR TITLE
CDRIVER-2611: Add server reply to mongoc_apm_command_failed_t

### PR DESCRIPTION
**Pull request by kvwalker**
JIRA ticket: https://jira.mongodb.org/browse/CDRIVER-2611
Evergreen patch: https://evergreen.mongodb.com/version/5ada3b45e3c33177f699637c

### CODE REVIEW

File: doc/mongoc_apm_command_failed_get_reply.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+Synopsis
+--------
+
+.. code-block:: c
+
+  const bson_t *
+  mongoc_apm_command_failed_get_reply (
+     const mongoc_apm_command_failed_t *event);
+
+Returns this event's reply. The data is only valid in the scope of the callback that receives this event; copy it if it will be accessed after the callback returns.
```
Comment (ajdavis) [inline doc/mongoc_apm_command_failed_get_reply.rst]: Please mention that this is the server's reply to a command that failed, it contains details about why the command failed. If no server reply was received (for example, if there was a network error) then the reply is a valid empty BSON document.

File: src/mongoc/mongoc-client.c [hunk 2]
```diff
@@ -2172,10 +2173,15 @@ _mongoc_client_monitor_op_killcursors_failed (
       EXIT;
    }
 
+   /* fake server reply to killCursors command: {ok: 0} */
+   bson_init (&doc);
+   bson_append_int32 (&doc, "ok", 2, 0);
```
Comment (ajdavis) [inline src/mongoc/mongoc-client.c]: bson_destroy the doc after executing the callback.

File: src/mongoc/mongoc-cursor.c [hunk 3]
```diff
@@ -737,10 +738,17 @@ _mongoc_cursor_monitor_failed (mongoc_cursor_t *cursor,
       EXIT;
    }
 
+    /* we sent OP_QUERY/OP_GETMORE, fake a reply to find/getMore command:
+    * {ok: 0}
+    */
```
Comment (ajdavis) [inline src/mongoc/mongoc-cursor.c]: destroy it later

File: src/mongoc/mongoc-client.c [hunk 4]
```diff
@@ -2185,6 +2191,7 @@ _mongoc_client_monitor_op_killcursors_failed (
    client->apm_callbacks.failed (&event);
 
    mongoc_apm_command_failed_cleanup (&event);
+   bson_destroy(&doc);
```
Comment (ajdavis) [inline src/mongoc/mongoc-client.c]: Space before paren - you can read the "C Driver Development" guide in Google docs if you want to see how to run clang-format, or just manually format the code according to our style.

File: src/mongoc/mongoc-cursor.c [hunk 5]
```diff
@@ -750,6 +758,7 @@ _mongoc_cursor_monitor_failed (mongoc_cursor_t *cursor,
    client->apm_callbacks.failed (&event);
 
    mongoc_apm_command_failed_cleanup (&event);
+   bson_destroy(&reply);
```
Comment (ajdavis) [inline src/mongoc/mongoc-cursor.c]: Space.

File: tests/test-mongoc-command-monitoring.c [hunk 6]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
...
+   mongoc_client_t *client;
+   mongoc_apm_callbacks_t *callbacks;
+   mongoc_collection_t *collection;
+   cmd_test_t test;
+   mongoc_cursor_t *cursor;
+   const bson_t *doc;
+   future_t *future;
+   request_t *request;
+
+   /* test that the command_failed_event's reply is the same as a mocked reply*/
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: space: "reply */"

File: tests/test-mongoc-command-monitoring.c [hunk 7]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
...
+   capture_logs (true);
+   mongoc_cursor_destroy (cursor);
+   mongoc_collection_destroy (collection);
+
+   mongoc_client_destroy (client);
+
+   cmd_test_cleanup (&test);
+   mongoc_apm_callbacks_destroy (callbacks);
+
+   /* test that the command_failed_event's reply is empty if there is a network error (i.e. the server hangs up) */
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: is this line over 80 chars?

File: tests/test-mongoc-command-monitoring.c [hunk 8]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
...
+   cmd_test_init (&test);
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: let's make this two test functions to test the two scenarios - starting here it seems you reinit all the variables anyway

File: tests/test-mongoc-command-monitoring.c [hunk 9]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
...
+
+   server = mock_server_with_autoismaster (4);
+   mock_server_run (server);
+
+   callbacks = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_command_started_cb(callbacks, command_failed_reply_command_started_cb);
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: space before paren

File: tests/test-mongoc-command-monitoring.c [hunk 10]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
 }
 
 
+static void
+command_failed_reply_command_started_cb (const mongoc_apm_command_started_t *event)
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: does it buy us anything to register a command-started callback? i think we're only testing the new feature below, in the command-failed callback.

File: tests/test-mongoc-command-monitoring.c [hunk 11]
```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
...
+   cmd_test_t *test;
+
+   test = (cmd_test_t *) mongoc_apm_command_failed_get_context (event);
+   test->failed_calls++;
+   bson_destroy (&test->cmd);
+
+   ASSERT_CMPSTR (test->cmd_name,
+                  mongoc_apm_command_failed_get_command_name (event));
+
+   bson_copy_to (mongoc_apm_command_failed_get_reply (event), &test->cmd);
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: it's confusing to copy "reply" to "cmd", i think it's better to *not* reuse the cmd_test_t struct in this test and instead make a new struct for the purpose of testing command-failed events' replies

File: tests/test-mongoc-command-monitoring.c (hunk 10 above)
Comment (kvwalker) [reply tests/test-mongoc-command-monitoring.c]: The only reason I used a command-started callback was to assert that the command_name was to assert that `test->cmd_name == mongoc_apm_command_failed_get_command_name (event)`. I removed that to remove the command-started callback.

File: tests/test-mongoc-command-monitoring.c [hunk 12]
```diff
@@ -1745,6 +1745,159 @@ test_killcursors_deprecated (void)
...
+
+   mongoc_client_destroy (client);
+
+   cmd_failed_reply_test_cleanup (&test);
+   mongoc_apm_callbacks_destroy (callbacks);
+}
+
+
+static void
+test_command_failed_reply_drop (void)
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: Call this "hangup" or "disconnect"

File: tests/test-mongoc-command-monitoring.c [hunk 13]
```diff
@@ -1745,6 +1745,159 @@ test_killcursors_deprecated (void)
 }
 
 
+typedef struct {
+   int failed_calls;
+   char db[100];
+   char cmd_name[100];
```
Comment (ajdavis) [inline tests/test-mongoc-command-monitoring.c]: Are all these fields used?

**Comment by ajdavis** [comment]
LGTM! Please squash to one commit with a message of at most 50 chars. C Driver commit messages follow this pattern:

CDRIVER-NNNN lowercase commit message

There's no colon after the ticket number, and no period at the end of the subject line.

**Comment by kvwalker** [comment]
Thank you for all your help!

**Pull request merged**